File: python/neighborhoods.py
from structures import *


@dataclass
class Move:
    data: Tuple


class Neighborhood(ABC):
    def __init__(self, I: Instance, sol: Solution):
        self.I = I
        self.sol = sol
        self.f = objective(self.I, self.sol)

    def generate(self) -> Iterator[Move]:
        raise RuntimeError("Call to abstract Class")

    def apply(self, mov: Move) -> Solution:
        raise RuntimeError("Call to abstract Class")

    def is_valid(self, mov: Move) -> bool:
        raise RuntimeError("Call to abstract Class")

    def calc_delta(self, mov: Move) -> float:
        raise RuntimeError("Call to abstract Class")


class IntraRouteNeighborhood(Neighborhood):
    """
    Apply permutation inside a single tracks route.
    """

    def generate(self) -> Iterator[Move]:
        for i, route in enumerate(self.sol.routes):
            m = len(route)
            for k in range(m):
                for l in range(k + 1, m):
                    
                    # The move is the index of the 
                    # route the first node and the second node
                    yield Move(data=(i, k, l))

    def apply(self, mov: Move) -> Solution:
        i, k, l = mov.data
        new_sol = deepcopy(self.sol)
        route = new_sol.routes[i]
        route[k], route[l] = route[l], route[k]

        return new_sol

    def is_valid(self, mov: Move) -> bool:
        i, k, l = mov.data
        route = copy(self.sol.routes[i])
        route[k], route[l] = route[l], route[k]
        cargo = calc_route_cargo(self.I, route)
        return bool(np.all(cargo < self.I.C))

    
    def calc_delta(self, mov: Move) -> float:
        r, k, l = mov.data
        route = self.sol.routes[r]
        I = self.I
        dist = I.dist

        # nodes before and after positions
        x = route[k]
        y = route[l]

        # A, B around x
        A = route[k - 1] if k > 0 else 0
        B = route[k + 1] if k + 1 < len(route) else 0

        # C, D around y
        C = route[l - 1] if l > 0 else 0
        D = route[l + 1] if l + 1 < len(route) else 0

        # special case: if swapping adjacent nodes k+1 = l
        # adjust arcs (A, x, y, D)
        if l == k + 1:
            # original: A → x → y → D
            # new:      A → y → x → D
            delta = (
                dist[A, y] + dist[y, x] + dist[x, D]
                - dist[A, x] - dist[x, y] - dist[y, D]
            )
            return delta

        # general case
        delta = (
            dist[A, y] + dist[y, B] + dist[C, x] + dist[x, D]
            - dist[A, x] - dist[x, B] - dist[C, y] - dist[y, D]
        )

        return delta



# ------------------------------------------------------------
# Helper: find pickup & delivery positions of requests in route
# ------------------------------------------------------------
def pickup_delivery_positions(I: Instance, route: Route):
    """
    Returns a list of (p_idx, d_idx, req_id, pickup_node, delivery_node)
    for all requests whose pickup & delivery are in this route.
    """
    pos = {}
    for idx, node in enumerate(route):
        req = I.request_of_node[node]
        if req >= 0:
            if req not in pos:
                pos[req] = [None, None, node]  # p_idx, d_idx, pickup_or_delivery_node
            # Check whether node is pickup or delivery
            if I.load_change[node] > 0:   # pickup
                pos[req][0] = idx
            else:                          # delivery
                pos[req][1] = idx

    out = []
    for req, (p_idx, d_idx, _) in pos.items():
        if p_idx is not None and d_idx is not None:
            pickup_node = 2 * req + 1
            delivery_node = 2 * req + 2
            out.append((p_idx, d_idx, req, pickup_node, delivery_node))
    return out


# =====================================================================
# 1. Pair-Relocate Neighborhood (inter-route)
# =====================================================================

class PairRelocateNeighborhood(Neighborhood):

    def generate(self) -> Iterator[Move]:
        """
        Moves one request (pickup+delivery) from route A to route B.
        """
        for r_from, routeA in enumerate(self.sol.routes):
            # find all pickup/delivery pairs
            for (p_old, d_old, req, pnode, dnode) in pickup_delivery_positions(self.I, routeA):

                for r_to, routeB in enumerate(self.sol.routes):
                    if r_to == r_from:
                        continue

                    # Try all possible new insertion positions such that p_new < d_new
                    for p_new in range(len(routeB) + 1):
                        for d_new in range(p_new + 1, len(routeB) + 2):
                            data = ("pair_relocate",
                                    r_from, p_old, d_old,
                                    r_to, p_new, d_new,
                                    req, pnode, dnode)
                            yield Move(data=data)

    # ----------------------------------------------------------

    def apply(self, mov: Move) -> Solution:
        tag, r_from, p_old, d_old, r_to, p_new, d_new, req, pnode, dnode = mov.data

        new_sol = deepcopy(self.sol)
        routeA = new_sol.routes[r_from]
        routeB = new_sol.routes[r_to]

        # Remove delivery first (higher index), then pickup.
        pA = routeA.pop(d_old)
        pB = routeA.pop(p_old)

        # Insert into route B
        routeB.insert(p_new, pnode)
        # delivery index shifts because of pickup insertion
        routeB.insert(d_new, dnode)

        return new_sol

    # ----------------------------------------------------------

    def is_valid(self, mov: Move) -> bool:
        tag, r_from, p_old, d_old, r_to, p_new, d_new, req, pnode, dnode = mov.data

        # We simulate small route changes
        routeA = self.sol.routes[r_from][:]
        routeB = self.sol.routes[r_to][:]

        # Apply the swap to these local copies:

        # Remove in descending order
        del routeA[d_old]
        del routeA[p_old]

        # Insert to route B
        routeB.insert(p_new, pnode)
        routeB.insert(d_new, dnode)

        # Capacity check on BOTH routes
        return bool(
            np.all(calc_route_cargo(self.I, routeA) <= self.I.C) and
            np.all(calc_route_cargo(self.I, routeB) <= self.I.C)
        )

    # ----------------------------------------------------------

    def calc_delta(self, mov: Move) -> float:
        tag, r_from, p_old, d_old, r_to, p_new, d_new, req, pnode, dnode = mov.data
        I = self.I
        dist = I.dist

        # Old routes
        routeA = self.sol.routes[r_from]
        routeB = self.sol.routes[r_to]

        # --------------------------------
        # ARCS REMOVED IN ROUTE A
        # --------------------------------

        # pickup removal
        A_prev = routeA[p_old - 1] if p_old > 0 else 0
        A_next = routeA[p_old + 1] if p_old + 1 < len(routeA) else 0

        # delivery removal: indices shift if d_old > p_old
        d_old_shift = d_old
        # If delivery appears after pickup, after removing pickup first,
        # its index decreases by 1. BUT we compute arc changes BEFORE removal.
        D_prev = routeA[d_old - 1] if d_old > 0 else 0
        D_next = routeA[d_old + 1] if d_old + 1 < len(routeA) else 0

        removed = (
            dist[A_prev, pnode] + dist[pnode, A_next] +
            dist[D_prev, dnode] + dist[dnode, D_next]
        )

        # --------------------------------
        # ARCS ADDED IN ROUTE A (after deletion)
        # Actually: pickup and delivery are removed completely,
        # so route A gains arc (A_prev -> A_next) at pickup position,
        # and (D_prev -> D_next) at delivery.
        # --------------------------------

        added = 0

        # For pickup gap:
        added += dist[A_prev, A_next]

        # For delivery gap:
        added += dist[D_prev, D_next]

        deltaA = added - removed

        # =================================================================
        # NOW route B: inserting pickup at p_new and delivery at d_new
        # =================================================================

        # After pickup insertion:
        B_prev_p = routeB[p_new - 1] if p_new > 0 else 0
        B_next_p = routeB[p_new] if p_new < len(routeB) else 0

        # Delivery insertion (after pickup inserted):
        B_prev_d = routeB[d_new - 1] if d_new > 0 else 0
        B_next_d = routeB[d_new] if d_new < len(routeB) else 0

        removedB = (
            dist[B_prev_p, B_next_p] +
            dist[B_prev_d, B_next_d]
        )

        addedB = (
            dist[B_prev_p, pnode] + dist[pnode, B_next_p] +
            dist[B_prev_d, dnode] + dist[dnode, B_next_d]
        )

        deltaB = addedB - removedB

        return deltaA + deltaB


# =====================================================================
# 2. Two-Opt Neighborhood (PD-compatible)
# =====================================================================

class TwoOptNeighborhood(Neighborhood):

    def generate(self) -> Iterator[Move]:
        for r_idx, route in enumerate(self.sol.routes):
            m = len(route)
            for i in range(m - 2):
                for j in range(i + 2, m):  # non-adjacent to avoid trivial swap
                    yield Move(("2opt", r_idx, i, j))

    # ----------------------------------------------------------

    def apply(self, mov: Move) -> Solution:
        tag, r_idx, i, j = mov.data
        new_sol = deepcopy(self.sol)
        route = new_sol.routes[r_idx]
        route[i:j+1] = reversed(route[i:j+1])
        return new_sol

    # ----------------------------------------------------------

    def is_valid(self, mov: Move) -> bool:
        tag, r_idx, i, j = mov.data
        route = self.sol.routes[r_idx][:]

        # Reverse segment
        route[i:j+1] = reversed(route[i:j+1])

        # Capacity
        if not np.all(calc_route_cargo(self.I, route) <= self.I.C):
            return False

        # Precedence: pickup must occur before delivery
        for idx, node in enumerate(route):
            req = self.I.request_of_node[node]
            if req >= 0:
                # pickup and delivery nodes
                pnode = 2 * req + 1
                dnode = 2 * req + 2
                if dnode in route and pnode in route:
                    if route.index(pnode) > route.index(dnode):
                        return False

        return True

    # ----------------------------------------------------------

    def calc_delta(self, mov: Move) -> float:
        """Delta evaluation for 2-opt: consider boundary arc changes."""
        tag, r_idx, i, j = mov.data
        dist = self.I.dist
        route = self.sol.routes[r_idx]

        # Edges removed: A->x, y->B
        A = route[i - 1] if i > 0 else 0
        x = route[i]
        y = route[j]
        B = route[j + 1] if j + 1 < len(route) else 0

        removed = dist[A, x] + dist[y, B]
        added   = dist[A, y] + dist[x, B]

        # Internal reversal cost changes require full scan
        old_internal = 0
        new_internal = 0

        seg = route[i:j+1]

        for u, v in zip(seg, seg[1:]):
            old_internal += dist[u, v]  # original

        rev = seg[::-1]
        for u, v in zip(rev, rev[1:]):
            new_internal += dist[u, v]  # reversed

        return (added - removed) + (new_internal - old_internal)

File: cpp/core/src/neighborhoods.cpp
#include <algorithm>
#include <numeric>
#include "neighborhoods.hpp"
#include "utils.hpp"
#include <iostream>
// Creates a vector of info for each request in a route.
std::vector<PickupDeliveryInfo>
pickup_delivery_positions(const Instance &I, const std::vector<int> &route)
{
    struct Pos
    {
        int p_idx = -1;
        int d_idx = -1;
    };
    std::unordered_map<int, Pos> pos;

    for (int idx = 0; idx < (int)route.size(); ++idx)
    {
        int node = route[idx];
        int req = I.request_of_node[node];
        if (req >= 0)
        {
            if (I.load_change[node] > 0)
                pos[req].p_idx = idx;
            else
                pos[req].d_idx = idx;
        }
    }

    std::vector<PickupDeliveryInfo> out;
    out.reserve(pos.size());
    for (const auto &kv : pos)
    {
        int req = kv.first;
        int p_idx = kv.second.p_idx;
        int d_idx = kv.second.d_idx;
        if (p_idx >= 0 && d_idx >= 0)
        {
            int pickup_node = 2 * req + 1;
            int delivery_node = 2 * req + 2;
            out.push_back(PickupDeliveryInfo{p_idx, d_idx, req, pickup_node, delivery_node});
        }
    }
    return out;
}

void IntraRouteNeighborhood::generate(std::vector<GenericMove> &moves) const
{
    moves.clear();
    for (int r = 0; r < (int)sol.routes.size(); ++r)
    {
        const auto &route = sol.routes[r];
        int m = (int)route.size();
        for (int k = 0; k < m; ++k)
        {
            for (int l = k + 1; l < m; ++l) // No for l<k+1 because of double counting.
            {
                moves.push_back(GenericMove{1, {r, k, l}});
            }
        }
    }
}

std::optional<GenericMove>
IntraRouteNeighborhood::generate_random(std::mt19937 &rng) const
{
    int R = (int)sol.routes.size();
    if (R == 0)
        return std::nullopt;

    std::uniform_int_distribution<int> route_dist(0, R - 1);

    for (int tries = 0; tries < this->MAX_TRIES_RANDOM; tries++)
    {
        // 1) Pick a random route
        int r = route_dist(rng);
        const auto &route = sol.routes[r];
        int m = (int)route.size();

        // Must have at least 2 positions
        if (m < 2)
            continue;

        std::uniform_int_distribution<int> idx_dist(0, m - 1);

        // 2) Pick random indices k < l
        int k = idx_dist(rng);
        int l = idx_dist(rng);
        if (k == l)
            continue;
        if (k > l)
            std::swap(k, l);

        // 3) Build the move
        GenericMove mvs{1, {r, k, l}};

        // 4) Validate
        if (is_valid(mvs))
            return mvs;
    }

    return std::nullopt;
}

bool IntraRouteNeighborhood::is_valid(const GenericMove &mov) const
{
    int r = mov.data[0];
    int k = mov.data[1];
    int l = mov.data[2];

    auto route = sol.routes[r];
    std::swap(route[k], route[l]);

    auto cargo = utils::calc_route_cargo(I, route);
    for (int c : cargo)
    {
        if (c >= I.C)
            return false;
    }
    return true;
}

double IntraRouteNeighborhood::calc_delta(const GenericMove &mov) const
{
    int r = mov.data[0];
    int k = mov.data[1];
    int l = mov.data[2];

    if (k == l)
        return 0.0;
    if (k > l)
        std::swap(k, l);

    const auto &route = sol.routes[r];
    const auto &dist = I.dist;

    int x = route[k];
    int y = route[l];

    int A = (k > 0) ? route[k - 1] : 0;
    int B = (k + 1 < (int)route.size()) ? route[k + 1] : 0;
    int C = (l > 0) ? route[l - 1] : 0;
    int D = (l + 1 < (int)route.size()) ? route[l + 1] : 0;

    int delta_d;

    if (l == k + 1)
    {
        delta_d =
            dist[A][y] + dist[y][x] + dist[x][D] -
            (dist[A][x] + dist[x][y] + dist[y][D]);
    }
    else
    {
        delta_d =
            dist[A][y] + dist[y][B] + dist[C][x] + dist[x][D] -
            (dist[A][x] + dist[x][B] + dist[C][y] + dist[y][D]);
    }

    double d_old = utils::route_distance(I, route);
    double d_new = d_old + delta_d;

    double S_old = sol.total_distance;
    double Q_old = sol.sum_of_squares;

    double S_new = S_old - d_old + d_new;
    double Q_new = Q_old - d_old * d_old + d_new * d_new;

    double J_old = (S_old * S_old) / (I.nK * Q_old);
    double J_new = (S_new * S_new) / (I.nK * Q_new);

    return delta_d + I.rho * (J_old - J_new);
}

Solution IntraRouteNeighborhood::apply(const GenericMove &mov) const
{
    int r = mov.data[0];
    int k = mov.data[1];
    int l = mov.data[2];

    Solution new_sol = sol;
    std::swap(new_sol.routes[r][k], new_sol.routes[r][l]);

    auto all_distances = utils::all_route_distances(I, new_sol);
    std::vector<double> sq_distances;
    sq_distances.resize(all_distances.size());
    std::transform(all_distances.begin(), all_distances.end(), sq_distances.begin(), [](auto val)
                   { return val * val; });
    new_sol.total_distance = std::accumulate(all_distances.begin(), all_distances.end(), 0.0);

    new_sol.sum_of_squares = std::accumulate(sq_distances.begin(), sq_distances.end(), 0.0);
    return new_sol;
}

// =====================================================================
// 2. PairRelocateNeighborhood
// =====================================================================

void PairRelocateNeighborhood::generate(std::vector<GenericMove> &moves) const
{
    moves.clear();
    int R = (int)sol.routes.size();

    for (int r_from = 0; r_from < R; ++r_from)
    {
        const auto &routeA = sol.routes[r_from];
        auto infos = pickup_delivery_positions(I, routeA);

        for (const auto &info : infos)
        {
            for (int r_to = 0; r_to < R; ++r_to)
            {
                if (r_to == r_from)
                    continue;

                int lenB = (int)sol.routes[r_to].size();
                for (int p_new = 0; p_new <= lenB; ++p_new)
                {
                    for (int d_new = p_new + 1; d_new <= lenB + 1; ++d_new)
                    {
                        moves.push_back(GenericMove{
                            2,
                            {r_from, info.p_idx, info.d_idx,
                             r_to, p_new, d_new,
                             info.req, info.pickup_node, info.delivery_node}});
                    }
                }
            }
        }
    }
}

std::optional<GenericMove>
PairRelocateNeighborhood::generate_random(std::mt19937 &rng) const
{
    int R = (int)sol.routes.size();
    if (R < 2)
        return std::nullopt;

    // 1) Pick r_from randomly
    std::uniform_int_distribution<int> route_dist(0, R - 1);

    for (int tries = 0; tries < this->MAX_TRIES_RANDOM; tries++)
    {
        int r_from = route_dist(rng);
        const auto &routeA = sol.routes[r_from];

        // Need at least one pickup-delivery pair
        auto infos = pickup_delivery_positions(I, routeA);
        if (infos.empty())
            continue;

        // 2) Choose a random pair (pickup_idx, delivery_idx)
        std::uniform_int_distribution<int> info_dist(0, (int)infos.size() - 1);
        const auto &info = infos[info_dist(rng)];

        // 3) Pick r_to randomly, different from r_from
        int r_to = r_from;
        if (R > 1)
        {
            do
            {
                r_to = route_dist(rng);
            } while (r_to == r_from);
        }

        const auto &routeB = sol.routes[r_to];
        int lenB = (int)routeB.size();
        if (lenB < 0)
            continue;

        // 4) Random insertion positions for pickup and delivery
        std::uniform_int_distribution<int> pos_dist(0, lenB + 1); // inclusive

        int p_new = pos_dist(rng);
        std::uniform_int_distribution<int> d_dist(p_new + 1, lenB + 2);
        int d_new = d_dist(rng);

        // 5) Build the candidate move
        GenericMove m{
            2,
            {r_from, info.p_idx, info.d_idx,
             r_to, p_new, d_new,
             info.req, info.pickup_node, info.delivery_node}};

        // 6) Validate it
        if (is_valid(m))
            return m;
    }

    return std::nullopt;
}

bool PairRelocateNeighborhood::is_valid(const GenericMove &mov) const
{
    int r_from = mov.data[0];
    int p_old = mov.data[1];
    int d_old = mov.data[2];
    int r_to = mov.data[3];
    int p_new = mov.data[4];
    int d_new = mov.data[5];
    int pnode = mov.data[7];
    int dnode = mov.data[8];

    auto routeA = sol.routes[r_from];
    auto routeB = sol.routes[r_to];

    // FIX 1: erase in correct order depending on indices
    if (p_old < d_old)
    {
        routeA.erase(routeA.begin() + d_old);
        routeA.erase(routeA.begin() + p_old);
    }
    else
    {
        routeA.erase(routeA.begin() + p_old);
        routeA.erase(routeA.begin() + d_old);
    }

    if (p_new < 0 || p_new > (int)routeB.size())
        return false;
    if (d_new < 0 || d_new > (int)routeB.size() + 1)
        return false;

    routeB.insert(routeB.begin() + p_new, pnode);
    // after inserting pickup, delivery index is in [p_new+1, size]
    if (d_new > (int)routeB.size())
        return false;
    routeB.insert(routeB.begin() + d_new, dnode);

    for (int c : utils::calc_route_cargo(I, routeA))
        if (c > I.C)
            return false;
    for (int c : utils::calc_route_cargo(I, routeB))
        if (c > I.C)
            return false;

    return true;
}

// double PairRelocateNeighborhood::calc_delta(const GenericMove &mov) const
// {
//     int r_from = mov.data[0];
//     int p_old = mov.data[1];
//     int d_old = mov.data[2];
//     int r_to = mov.data[3];
//     int p_new = mov.data[4];
//     int d_new = mov.data[5];
//     int pnode = mov.data[7];
//     int dnode = mov.data[8];

//     const auto &routeA_orig = sol.routes[r_from];
//     const auto &routeB_orig = sol.routes[r_to];

//     // old total distance of the two routes
//     double old_cost =
//         (double)utils::route_distance(I, routeA_orig) +
//         (double)utils::route_distance(I, routeB_orig);

//     // apply move to copies
//     auto routeA = routeA_orig;
//     auto routeB = routeB_orig;

//     if (p_old < d_old)
//     {
//         routeA.erase(routeA.begin() + d_old);
//         routeA.erase(routeA.begin() + p_old);
//     }
//     else
//     {
//         routeA.erase(routeA.begin() + p_old);
//         routeA.erase(routeA.begin() + d_old);
//     }

//     routeB.insert(routeB.begin() + p_new, pnode);
//     if (d_new > (int)routeB.size())
//     {
//         // if out of bound, treat as worst-case big delta
//         return std::numeric_limits<double>::infinity();
//     }
//     routeB.insert(routeB.begin() + d_new, dnode);

//     double new_cost =
//         (double)utils::route_distance(I, routeA) +
//         (double)utils::route_distance(I, routeB);

//     return new_cost - old_cost;
// }

Solution PairRelocateNeighborhood::apply(const GenericMove &mov) const
{
    int r_from = mov.data[0];
    int p_old = mov.data[1];
    int d_old = mov.data[2];
    int r_to = mov.data[3];
    int p_new = mov.data[4];
    int d_new = mov.data[5];
    int pnode = mov.data[7];
    int dnode = mov.data[8];

    Solution new_sol = sol;

    auto &routeA = new_sol.routes[r_from];
    auto &routeB = new_sol.routes[r_to];

    if (p_old < d_old)
    {
        routeA.erase(routeA.begin() + d_old);
        routeA.erase(routeA.begin() + p_old);
    }
    else
    {
        routeA.erase(routeA.begin() + p_old);
        routeA.erase(routeA.begin() + d_old);
    }

    routeB.insert(routeB.begin() + p_new, pnode);
    routeB.insert(routeB.begin() + d_new, dnode);

    auto all_distances = utils::all_route_distances(I, new_sol);
    std::vector<double> sq_distances;
    sq_distances.resize(all_distances.size());
    std::transform(all_distances.begin(), all_distances.end(), sq_distances.begin(), [](auto val)
                   { return val * val; });
    new_sol.total_distance = std::accumulate(all_distances.begin(), all_distances.end(), 0.0);

    new_sol.sum_of_squares = std::accumulate(sq_distances.begin(), sq_distances.end(), 0.0);

    return new_sol;
}
double PairRelocateNeighborhood::calc_delta(const GenericMove &mov) const
{
    int r_from = mov.data[0];
    int p_old  = mov.data[1];
    int d_old  = mov.data[2];
    int r_to   = mov.data[3];
    int p_new  = mov.data[4];
    int d_new  = mov.data[5];
    int pnode  = mov.data[7];
    int dnode  = mov.data[8];

    const auto &routeA_orig = sol.routes[r_from];
    const auto &routeB_orig = sol.routes[r_to];

    double dA_old = utils::route_distance(I, routeA_orig);
    double dB_old = utils::route_distance(I, routeB_orig);

    auto routeA = routeA_orig;
    auto routeB = routeB_orig;

    if (p_old < d_old)
    {
        routeA.erase(routeA.begin() + d_old);
        routeA.erase(routeA.begin() + p_old);
    }
    else
    {
        routeA.erase(routeA.begin() + p_old);
        routeA.erase(routeA.begin() + d_old);
    }

    routeB.insert(routeB.begin() + p_new, pnode);
    if (d_new > (int)routeB.size())
        return std::numeric_limits<double>::infinity();
    routeB.insert(routeB.begin() + d_new, dnode);

    double dA_new = utils::route_distance(I, routeA);
    double dB_new = utils::route_distance(I, routeB);

    double S_old = sol.total_distance;
    double Q_old = sol.sum_of_squares;

    double S_new = S_old - dA_old - dB_old + dA_new + dB_new;
    double Q_new = Q_old
                 - dA_old * dA_old + dA_new * dA_new
                 - dB_old * dB_old + dB_new * dB_new;

    double J_old = (S_old * S_old) / (I.nK * Q_old);
    double J_new = (S_new * S_new) / (I.nK * Q_new);

    double delta_d = (dA_new - dA_old) + (dB_new - dB_old);
    return delta_d + I.rho * (J_old - J_new);
}



// =====================================================================
// 3. TwoOptNeighborhood
// =====================================================================

void TwoOptNeighborhood::generate(std::vector<GenericMove> &moves) const
{
    moves.clear();
    for (int r = 0; r < (int)sol.routes.size(); ++r)
    {
        const auto &route = sol.routes[r];
        int m = (int)route.size();
        for (int i = 0; i < m - 2; ++i)
        {
            for (int j = i + 2; j < m; ++j)
            {
                moves.push_back(GenericMove{3, {r, i, j}});
            }
        }
    }
}

std::optional<GenericMove>
TwoOptNeighborhood::generate_random(std::mt19937 &rng) const
{
    if (sol.routes.empty())
        return std::nullopt;

    std::uniform_int_distribution<int> route_dist(0, sol.routes.size() - 1);
    int rid = route_dist(rng);

    const auto &r = sol.routes[rid];
    if (r.size() < 4)
        return std::nullopt;

    std::uniform_int_distribution<int> idx_dist(0, r.size() - 1);

    for (int t = 0; t < this->MAX_TRIES_RANDOM; t++)
    { // few random attempts
        int i = idx_dist(rng);
        int j = idx_dist(rng);
        if (i == j)
            continue;
        if (i > j)
            std::swap(i, j);

        GenericMove m{3, {rid, i, j}};

        if (is_valid(m))
            return m;
    }

    return std::nullopt;
}

bool TwoOptNeighborhood::is_valid(const GenericMove &mov) const
{
    int r = mov.data[0];
    int i = mov.data[1];
    int j = mov.data[2];

    auto route = sol.routes[r];
    std::reverse(route.begin() + i, route.begin() + j + 1);

    for (int c : utils::calc_route_cargo(I, route))
        if (c > I.C)
            return false;

    int nreq = I.n;
    std::vector<int> p(nreq, -1), d(nreq, -1);

    for (int idx = 0; idx < (int)route.size(); ++idx)
    {
        int node = route[idx];
        int req = I.request_of_node[node];
        if (req < 0)
            continue;

        if (I.load_change[node] > 0)
        {
            if (p[req] == -1)
                p[req] = idx;
        }
        else
        {
            if (d[req] == -1)
                d[req] = idx;
        }
    }

    for (int req = 0; req < nreq; ++req)
    {
        if (p[req] != -1 && d[req] != -1 && p[req] > d[req])
            return false;
    }

    return true;
}

double TwoOptNeighborhood::calc_delta(const GenericMove &mov) const
{
    int r = mov.data[0];
    int i = mov.data[1];
    int j = mov.data[2];

    const auto &route = sol.routes[r];
    const auto &dist = I.dist;

    int A = (i > 0) ? route[i - 1] : 0;
    int x = route[i];
    int y = route[j];
    int B = (j + 1 < (int)route.size()) ? route[j + 1] : 0;

    int removed = dist[A][x] + dist[y][B];
    int added = dist[A][y] + dist[x][B];

    double delta_d = added - removed;

    double d_old = utils::route_distance(I, route);
    double d_new = d_old + delta_d;

    double S_old = sol.total_distance;
    double Q_old = sol.sum_of_squares;

    double S_new = S_old - d_old + d_new;
    double Q_new = Q_old - d_old * d_old + d_new * d_new;

    double J_old = (S_old * S_old) / (I.nK * Q_old);
    double J_new = (S_new * S_new) / (I.nK * Q_new);

    return delta_d + I.rho * (J_old - J_new);
}

Solution TwoOptNeighborhood::apply(const GenericMove &mov) const
{
    int r = mov.data[0];
    int i = mov.data[1];
    int j = mov.data[2];

    Solution new_sol = sol;
    std::reverse(new_sol.routes[r].begin() + i, new_sol.routes[r].begin() + j + 1);

    auto all_distances = utils::all_route_distances(I, new_sol);
    std::vector<double> sq_distances;
    sq_distances.resize(all_distances.size());
    std::transform(all_distances.begin(), all_distances.end(), sq_distances.begin(), [](auto val)
                   { return val * val; });
    new_sol.total_distance = std::accumulate(all_distances.begin(), all_distances.end(), 0.0);

    new_sol.sum_of_squares = std::accumulate(sq_distances.begin(), sq_distances.end(), 0.0);

    return new_sol;
}
